### PR TITLE
Fix LUDCL optimization

### PR DESCRIPTION
**Pull request by theresa-m**
Fixes: https://github.com/eclipse/openj9/issues/7332
Doc issue: https://github.com/eclipse/openj9-docs/issues/413

Ports:
ibmruntimes/openj9-openjdk-jdk11#226
ibmruntimes/openj9-openjdk-jdk13#25
ibmruntimes/openj9-openjdk-jdk#149

**Summary:**
- The assumption that the ludcl will not change during a call to readObject/readUnshared can only be made if user code is not invoked. This change triggers a refresh of the cachedLudcl when a user method is invoked.
- Re-enable ludcl caching by default (com.ibm.enableClassCaching)

**Performance assessment:** (pending)
- Daytrader7: as a general benchmark for OpenJ9
- second benchmark that to more clearly show benchmark benefits: I am working with the perf team to get results for this. From previous documentation the patches in ObjectInputStream were critical to its performance

Signed-off-by: Theresa Mammarella <Theresa.T.Mammarella@ibm.com>

### CODE REVIEW

File: jdk/src/share/classes/java/io/ObjectInputStream.java [hunk 1]
```diff
@@ -337,26 +337,28 @@ public void checkArray(ObjectInputStream stream, Class<?> arrayType, int arrayLe
...
+    implements PrivilegedAction<Boolean> {
+        public Boolean run() {
+            String property =
+                System.getProperty("com.ibm.enableClassCaching", "true");
+            return property.equalsIgnoreCase("true");
+        }
+    }
     private ClassLoader cachedLudcl;
+    /* If user code is invoked in the middle of a call to readObject the cachedLudcl
+     * must be refreshed as the ludcl could have been changed while in user code. */
```
Comment (pshipton) [inline jdk/src/share/classes/java/io/ObjectInputStream.java]: For a multi-line comment, close the comment on the next line.

**Comment by pshipton** [comment]
Note this requires a doc PR as well.

**Comment by DanHeidinga** [comment]
> pending: performance impact assessment

+1.  Can you expand a bit on what will be tested to validate the performance of this change?

**Comment by theresa-m** [comment]
> > pending: performance impact assessment
> 
> +1. Can you expand a bit on what will be tested to validate the performance of this change?

- Daytrader7: as a general benchmark for OpenJ9
- SPECjbb2015: I am working with the perf team to get results for this. From previous documentation the patches in ObjectInputStream were critical to its performance

**Comment by theresa-m** [comment]
These are my performance results. The "without changes" builds have the optimization enabled by default.

Daytrader7:

  | Startup mean | Throughput mean | Throughput Warm mean
-- | -- | -- | --
With Changes | 4456 | 5578 | 5797
Without Changes | 4512 | 5656 | 5570
% difference | 1.25 | 1.40 | -3.92

SpecJbb2015 ppc:

  | Max-jops mean | Crit-jops mean
-- | -- | --
With changes | 136650 | 46173
Without changes | 140096 | 47778
% difference | 2.52 | 3.47

**Comment by zl-wang** [comment]
For consistency, only "Startup" is lower meaning better; all "Throughput" is lower meaning worse (regressed). Please mark the + and - % accordingly.

**Comment by vijaysun-omr** [comment]
Can we also get a measurement with the build with the LUDCL optimization disabled ?

**Comment by vijaysun-omr** [comment]
Is that "Throughput Warm mean" the warmed up shared classes cache run ?

**Comment by theresa-m** [comment]
thats the "LargeThreadPoolWarm" job

**Comment by theresa-m** [comment]
Sure thing. Ive added the CIs

**Comment by piyush286** [comment]
image

For more detailed results, please see this spreadsheet: https://ibm.box.com/s/sztz4lmb3urxvkjujqnn5wzrjv0mu1eu

There was one outlier in each of the sets that showed high CI. In the spreadsheet, I've marked the outliers in red. As you can in the above screenshot, I've added a column for `Without Outlier`, reflecting a more accurate picture. 

@theresa-m Hopefully, I haven't made any calculation mistakes, but feel free to double check. Btw it might be good to look at CIs for SPECjbb2015 runs as well.

**Comment by theresa-m** [comment]
Here are more results with the second benchmark, including with the ludcl optimization disabled:

  | Max-jops mean | Crit-jops mean
-- | -- | --
1 With Changes + LUDCL Enabled | 136650 | 46173
2 Without Changes + LUDCL Enabled | 140096 | 47778
3 LUDCL Disabled | 131298 | 45036
1 vs 2 (%) | 97.54 | 96.64
1 vs 3 (%) | 104.08 | 102.53
2 vs 3 (%) | 106.70 | 106.09

[ludcl_benchmark2.xlsx](https://github.com/ibmruntimes/openj9-openjdk-jdk8/files/3804377/ludcl_benchmark2.xlsx)

**Comment by DanHeidinga** [comment]
Summarizing the results so far:
* Theresa's fix makes the LUDCL optimization correct.
* It is an improvement over having the LUDCL optimization disabled.
* It's not quite as good as the broken version but being correct is better :)

This patch is worth delivering to restore most of the optimization.

Anyone want to weight in the above summary?
@theresa-m @vijaysun-omr @zl-wang @piyush286

**Comment by vijaysun-omr** [comment]
I agree that getting this change merged is an improvement over the current state, i.e. LUDCL optimization disabled. So please go ahead and do so.

**Comment by jamezp** [comment]
Just as an FYI I built with this patch and ran the test failing in https://github.com/eclipse/openj9/issues/7332 and it seems to work fine.

**Comment by DanHeidinga** [comment]
@theresa-m can the WIP be removed from this?

@pshipton Any other review comments or are you good with this?

**Comment by theresa-m** [comment]
I have removed the WIP.

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk8

**Comment by DanHeidinga** [comment]
@theresa-m Does this need to be ported to the other extension repos?  ie: 11, 13 & jdk?

**Comment by theresa-m** [comment]
> @theresa-m Does this need to be ported to the other extension repos? ie: 11, 13 & jdk?

Yes. I've just updated the description with the 11/13/next pull requests.

**Pull request merged**